internal/service/projects/main.go

```diff
@@ -0,0 +1,54 @@
+package projects
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/ucl-arc-tre/portal/internal/graceful"
+	"github.com/ucl-arc-tre/portal/internal/rbac"
+	"github.com/ucl-arc-tre/portal/internal/types"
+	"gorm.io/gorm"
+)
+
+type Service struct {
+	db *gorm.DB
+}
+
+func New() *Service {
+	return &Service{db: graceful.NewDB()}
+}
+
+// ValidateRoleAssignment checks if a user meets the requirements for a role
+// Example: To be a desktop user, you need to be an approved researcher (portal role)
+func (s *Service) ValidateRoleAssignment(userID uuid.UUID, roleName string) error {
+	// Get the user to check portal-level roles
+	var user types.User
+	if err := s.db.First(&user, userID).Error; err != nil {
+		return fmt.Errorf("user not found: %w", err)
+	}
+
+	// Role validation based on requirements
+	switch roleName {
+	case types.ProjectRoleIngresser:
+		// No requirements
+		return nil
```

Comment:
This is just an example to demonstrate the idea. The exact implementation could take alternative forms (e.g. a function with some sort of lookup).